### PR TITLE
Add overflow safe abs and enable UBSAN on travis

### PR DESCRIPTION
**Pull request by D4N**
This PR adds a new function `Safe::abs()` which, in contrast to `std::abs`, never invokes undefined behavior. It circumvents the issue of `abs(INTMIN) < 0` by returning `INTMAX` instead (which is not the exact result mathematically as `-1 * INTMAX == INTMIN + 1`.

This should allow us to enable UBSAN on travis & make it fatal.

### CODE REVIEW

**Review by piponazo — COMMENTED** [review]
In principle I am fine with the change if it is really fixing an existing issue or preventing a problem we could have in the future. In `floatToRationalCast` would not take such conflicting values, I would vote to keep the std version of abs.

File: src/types.cpp [hunk 1]
```diff
@@ -657,9 +660,10 @@ namespace Exiv2 {
     {
         // Beware: primitive conversion algorithm
         int32_t den = 1000000;
-        if (std::labs(static_cast<long>(f)) > 2147) den = 10000;
-        if (std::labs(static_cast<long>(f)) > 214748) den = 100;
-        if (std::labs(static_cast<long>(f)) > 21474836) den = 1;
+        const long f_as_long = static_cast<long>(f);
+        if (Safe::abs(f_as_long) > 2147) den = 10000;
```
[the remark this answers is not in the document]
Comment (D4N) [reply src/types.cpp]: This issue is already present in the test suite. If you compile exiv2 with `-fsanitize=undefined` and run `make tests` you'll get errors like this: 
```
Running stringto-test.sh ...
/home/dan/projects/exiv2/src/types.cpp:663:22: runtime error: negation of -9223372036854775808 cannot be represented in type 'long int'; cast to an unsigned type to negate this value to itself
/home/dan/projects/exiv2/src/types.cpp:664:22: runtime error: negation of -9223372036854775808 cannot be represented in type 'long int'; cast to an unsigned type to negate this value to itself
/home/dan/projects/exiv2/src/types.cpp:665:22: runtime error: negation of -9223372036854775808 cannot be represented in type 'long int'; cast to an unsigned type to negate this value to itself
```
This is also the output with gcc, with clang you'll get a different runtime error (with `-inf` or `inf`, depending on the clang version), as this is undefined behavior.

I agree that this is a corner case and fixing it makes no difference for the test suite, but I'd like to have UBSAN enabled on Travis and this is one of the issues that keep the testsuite from passing under UBSAN.

File: src/preview.cpp [hunk 2]
```diff
@@ -1038,7 +1038,9 @@ namespace Exiv2 {
     {
         pData_ = data.pData_;
         size_ = data.size_;
-        data.release();
+        // release the other DataBuf's buffers, but don't free them!
+        const std::pair<byte*, long> unused = data.release();
+        (void)(unused);
```
[the remark this answers is not in the document]
Comment (D4N) [reply src/preview.cpp]: Right, I forgot about that. I'll change it.

**Review by piponazo — APPROVED** [review]

File: .travis.yml [hunk 3]
```diff
@@ -6,20 +6,51 @@ matrix:
...
+      dist: trusty
+      sudo: required
+      compiler: gcc-8
+      addons:
+        apt:
+          update: true
+          sources:
+          - sourceline: 'ppa:ubuntu-toolchain-r/test'
+          packages:
+            - g++-8
```
Comment (piponazo) [inline .travis.yml]: Ey @D4N, I think it would be better to isolate this change about CI in another different PR. Or do you have a strong reason to do it also in this PR?

**Comment by piponazo** [comment]
Interestingly the clang jobs are failing with the following message when linking the **unit_tests** target:

```
/home/travis/conanData/gtest/1.8.0/bincrafters/stable/package/b3b29979865f60bae84bcbc17ff7fcf0fbe4d248/lib/libgtest.so: undefined reference to `std::__throw_out_of_range_fmt(char const*, ...)@GLIBCXX_3.4.20'
clang-5.0: error: linker command failed with exit code 1 (use -v to see invocation)
````
Do you have any idea why this could be happening?

File: .travis.yml (hunk 3 above)
Comment (D4N) [reply .travis.yml]: Yes, gcc 4.8 does not ship with UBSAN and clang's UBSAN detects different kinds of undefined behavior (the `getEnv` stuff was only detected by clang).

**Comment by D4N** [comment]
So, I managed to break Travis... It won't create any builds any more. I'll try to resubmit this.
